kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -166,16 +170,17 @@ public void produce(
             request.getTimestamp().orElse(Instant.now()));
 
     return produceResult
-        .handle(
+        .handleAsync(
             (result, error) -> {
               if (error != null) {
                 long latency = Duration.between(requestInstant, Instant.now()).toMillis();
                 recordErrorMetrics(latency);
                 throw new CompletionException(error);
               }
               return result;
-            })
-        .thenApply(
+            },
+            executorService)
+        .thenApplyAsync(
```

Comment:
No need to use a separate executor for the `apply`. Whoever closes the previous future (which is already a thread in `executorService` will execute the `apply` here.

Reply:
I ran one of the unit tests with the debugger on, with Executors.newFixedThreadPool(3) set as the executor service and saw that in some cases the same thread was used for both handle and thenApply, but not always:

handleAsync:
![image](https://user-images.githubusercontent.com/87994408/142191593-58d0b92a-a6e8-412e-b31b-b5a1d160463c.png)


thenApply:
![image](https://user-images.githubusercontent.com/87994408/142191650-d3b77e14-48bd-4559-b02c-aab47b815098.png)


or using thenApplyAsycn I see pool-1-thread-1 for the handleAsync and then this for the thenApply

![image](https://user-images.githubusercontent.com/87994408/142191908-70a45cfb-9c34-4ee0-af78-d5f4febcdf0c.png)

So, I think we need to explicitly specify the executor for both the handle and the thenApply.

The thread don't always match, but they are always from our thread pool

Reply:
The last picture seems very much expected - `thenApplyAsync` without the executor param will run on the common FJ pool.
Now on to the second picture, I am interested to see the lower part of the stack that reaches into the `thenApply` function. The `thenApply` applies to the result of `handle`, which is a new `CompletableFuture`. The contents of `handle` need to be executed (on the custom executor) in order to complete with the result that new `CompletableFuture`, and I am curious what reschedules on the `main` thread (which I assume is the one completing `controller.produce()`).

In order not to fret over one redundant context switch per response, we can keep the executor params on both right now, but I think we should:
1. collapse the `thenApply` into the `handle`.
2. examine the rescheduling on `main`.

Reply:
Yup I agree on the third picture, just included it for completeness.

I did think when I started this about collapsing the thenApply into the handle as I couldn't see anything async in the body of the handleAsync, and it just returns the result passed in anyway.

I decided against it as it always seems best to touch things as little as possible just before we are trying to release a product, but maybe that's the way forward.

handleAsync
![image](https://user-images.githubusercontent.com/87994408/142198657-1faa4a49-3007-42bc-a8e6-c52c231c7317.png)

thenApply (no executor)
![image](https://user-images.githubusercontent.com/87994408/142198712-13e1d436-1bd8-4dfa-9010-8f8a2f44ab2f.png)

Reply:
> I decided against it as it always seems best to touch things as little as possible just before we are trying to release a product, but maybe that's the way forward.

Absolutely!